Nextjs documentation update (#239)
tigrisdata/tigris-docs: docs/guides/nextjs/add-existing-app/index.mdx

import CodeBlock from "@theme/CodeBlock";
import createNextAppOutput from "!!raw-loader!./_createNextAppOutput";
import runDevOutput from "!!raw-loader!./_runDevOutput";
import loginOutput from "!!raw-loader!./_loginOutput";
import tigrisCreateAppOutput from "!!raw-loader!./_tigrisCreateAppOutput";
import Installation from "../../../includes/_install_cloud_login.mdx";

# Add Tigris to an existing Nextjs Application
This guide discusses the concepts behind using Tigris in your existing Nextjs project and configuring your application to integrate with Tigrisdata.

### Setup Tigris account
First, install the Tigris CLI with the command `brew install tigrisdata/tigris/tigris-cli` for macOS. For other operating systems: [See installation instructions here](https://docs.tigrisdata.com/cli/installation)

Next, sign in to your existing Tigris account or signup for a new account:
```shell
tigris login
```

```mdx-code-block
<TerminalWindow>
<CodeBlock language="shell">{loginOutput}</CodeBlock>
</TerminalWindow>
```

### Add Tigris Cloud Credentials

For your application to connect with Tigris, you need to generate some credentials and
add them to the environment file.

```shell

> tigris create application <YOUR_PROJECT_NAME> "<YOUR_PROJECT_DESCRIPTION>"

Reviewer: ```suggestion
tigris create application <YOUR_APP_NAME> "<YOUR_APP_DESCRIPTION>"
```

```

> Replace `<YOUR_PROJECT_NAME>` and `<YOUR_PROJECT_DESCRIPTION>` arguments with your project name and description, respectively. Then this will create the credentials and print them on the terminal screen:

Reviewer: ```suggestion
Replace `<YOUR_APP_NAME>` and `<YOUR_APP_DESCRIPTION>` arguments with your application's name and description, respectively. Then this will create the credentials and print them on the terminal screen:
```

```mdx-code-block
<TerminalWindow>
<CodeBlock language="shell">{tigrisCreateAppOutput}</CodeBlock>
</TerminalWindow>
```

You will use the `id` and `secret` from the output and add them as `TIGRIS_CLIENT_ID` and `TIGRIS_CLIENT_SECRET`
respectively in the file `.env.development.local` in your project.

Now, create a `.env.development.local`file in your project root directory and add the application credentials generated above to it:

```

> dotenv title=.env.development.local

Reviewer 2: Did you mean to have this line with the backticks above?

Author: No

> ```
> dotenv title=.env.development.local

Reviewer: ```suggestion
```dotenv title=.env.development.local
```

TIGRIS_URI=api.preview.tigrisdata.cloud
TIGRIS_CLIENT_ID=client_id_here

> TIGRIS

Reviewer 2: This is incomplete

```

### Define Models{#data-model}

Tigris stores data records as documents. Documents are analogous to JSON objects grouped in

> a [Collection](/overview/databases). Tigris follows a convention of having models and schemas in `models/tigris` directory. Within the directory, you'll need another folder that matches the name of your database name. Then create your schema file. 

Reviewer: ```suggestion
a [Collection](/overview/databases). Tigris follows a convention of having models and schemas in `models/tigris` directory. Within the directory, you'll need another folder that matches the name of your database. Then create your schema file. 
```

> Your model's structure should look like `models/tigris/[DATABASE_NAME]/[MODEL_NAME].tsx`

Reviewer: nit: just to be consistent
```suggestion
Your model's structure should look like `models/tigris/<DATABASE_NAME>/<MODEL_NAME>.tsx`
```

For example, create a `models/tigris/blogApp/blogItems.ts` file in the root directory of your project and add the code:

```ts title=models/tigris/todoApp/blogItems.ts
import {
  TigrisCollectionType,
  TigrisDataTypes,
  TigrisSchema,
} from "@tigrisdata/core/dist/types";

export const COLLECTION_NAME = "blogItems";

// highlight-start
export interface BlogItem extends TigrisCollectionType {

>     id?: number;

Reviewer 2: Don't make this field optional, remove the "?"

    title: string;
    content: string;
    description: string
}

export const BlogItemSchema: TigrisSchema<BlogItem> = {
  id: {
    type: TigrisDataTypes.INT32,
    primary_key: { order: 1, autoGenerate: true },
  },
  title: { type: TigrisDataTypes.STRING },
  content: { type: TigrisDataTypes.STRING },
  description: { type: TigrisDataTypes.STRING },
};
// highlight-end
```
You've imported the Tigris `TigrisCollectionType` type, which is extended to define a model for your collections, the `TigrisDataTypes` type that gives you access to the Tigris data types, and the `TigrisSchema` which is used to define the schema for collections.

### Tigris Automatic schema management

Now automate your schema management to configure how the Tigris instance loader will be executed at runtime and register your schema. First, create a `sripts/setup.ts` file in the root directory of your project and add the code:

```ts title=scripts/setup.ts
import { Tigris } from "@tigrisdata/core";
import { loadEnvConfig } from "@next/env";

> // Run the config loader only when not executing within next runtime

Reviewer 2: @adilansari do we have a simplified version of this or is the env stuff still needed?

Reviewer: Not yet. I was thinking to wait until 
1. we introduce project environments with branching and then a `export PROJECT_ENV="production"` is required that can be used to load all these variables internally in SDK
2. Decorators are available as an option to define schema

And then revisit this bootstrap altogether.

Reviewer 2: Sounds good

if (process.env.NODE_ENV === undefined) {
  if (process.env.APP_ENV === "development") {
    loadEnvConfig(process.cwd(), true);
  } else if (process.env.APP_ENV === "production") {
    loadEnvConfig(process.cwd());
  }
}

async function main() {

>   // highlight-start

Reviewer 2: Let's remove the highlights since the user needs to paste all the code here in their file

:::

### Create CRUD Operations

You can now create and execute CRUD operations against the `blogItems` collection. These
endpoints will be used by the client component of your application to perform CRUD operations in Tigris. Create a new folder in the `pages/api/` folder to 
define your API endpoints.

#### ⌲ GET `/api/blogs` to fetch all blogs

For this demonstration, create a  `pages/api/blogs/index.ts` file which will map to `/api/blogs` and add the code:

```ts title=pages/api/blogs/index.ts
import {
  COLLECTION_NAME,
  BlogItem,
} from '../../../models/tigris/blogApp/blogtems'

async function handleGet(req: NextApiRequest, res: NextApiResponse<Response>) {
  // highlight-start
  try {
    const itemsCollection = tigrisDb.getCollection<BlogItem>(COLLECTION_NAME);
    const cursor = itemsCollection.findMany();
    const items = await cursor.toArray();
    res.status(200).json({ result: items });
  } catch (err) {
    const error = err as Error;
    res.status(500).json({ error: error.message });
  }
  // highlight-end
}
```

Here we used a simple `findMany()` construct to read documents from a Tigris collection and returned
them as json.

> #### ⌲ POST `/api/items` to add to-do item

Reviewer 2: add blog item

> Now, create a `POST` method in `pages/api/items/index.ts` that will insert an item into this collection.

Reviewer 2: Please fix all such occurrences of `pages/api/items/index.ts` in this guide.

```ts title=pages/api/blogs/index.ts

async function handlePost(req: NextApiRequest, res: NextApiResponse<Response>) {
  // highlight-start
  try {
    const item = JSON.parse(req.body) as BlogItem;
    const itemsCollection = tigrisDb.getCollection<BlogItem>(COLLECTION_NAME);
    const inserted = await itemsCollection.insertOne(item);
    res.status(200).json({ result: [inserted] });
  } catch (err) {
    const error = err as Error;
    res.status(500).json({ error: error.message });
  }
  // highlight-end
}
```

Here we implemented a function that enables the user to insert a new item into the collection list using the `insertOne()` constructor.

> #### ⌲ PUT `/api/item/[id]` to update a record in the collection.

Reviewer 2: to update the contents of the blog item

Reviewer 2: The path should be `/api/blog/[id]`. Can you fix the rest of the tutorial as well

To update the records in the collection, create an add a `pages/api/item/[id].ts` file to add the `PUT` method, which is routed to the `api/item/[id] endpoint, where `id` is the unique
identifier for the item that is being updated:

```ts title=pages/api/item/[id].ts
async function handlePut(req: NextApiRequest, res: NextApiResponse<Data>) {
  // highlight-start
  try {
    const item = JSON.parse(req.body) as BlogItem;
    const itemsCollection = tigrisDb.getCollection<BlogItem>(COLLECTION_NAME);
    const updated = await itemsCollection.insertOrReplaceOne(item);
    res.status(200).json({ result: updated });
  } catch (err) {
    const error = err as Error;
    res.status(500).json({ error: error.message });
  }
  // highlight-end
}
```

We are leveraging the `insertOrReplaceOne()` function offered by the Tigris SDK that will entirely replace the item with the new one. You can as well use the [`updateOne()`](/typescript/documents/update)
the function here is to update individual fields instead of replacing the whole document.

> #### ⌲ DELETE `/api/item/[id]` to remove item from collection.

Reviewer 2: to remove the blog item

Reviewer 2: And `/api/blog/[id]`

Similar to the PUT method above, we will now add a `DELETE` method to the `pages/api/item/[id].ts` file to
remove items from your collection.

```ts title=pages/api/item/[id].ts
async function handleDelete(
  req: NextApiRequest,
  res: NextApiResponse<Data>,
  itemId: number
) {
  // highlight-start
  try {
    const itemsCollection = tigrisDb.getCollection<TodoItem>(COLLECTION_NAME);
    const status = (await itemsCollection.deleteOne({ id: itemId })).status;
    if (status === "deleted") {
      res.status(200).json({});
    } else {
      res.status(500).json({ error: `Failed to delete ${itemId}` });
    }
  } catch (err) {
    const error = err as Error;
    res.status(500).json({ error: error.message });
  }
  // highlight-end
}
```

The `deleteOne()` function accepts a filter and limits the deletion to only one document in the collection
matching the given filter. The operation has no effect if no document matches the specified filter.

> #### ⌲ GET `/api/items/search` to find items matching the given query

Reviewer 2: to find blog items matching the search query

Reviewer 2: And `/api/blogs/search`

> The search endpoint will allow users to find items in the list. Let's add this endpoint that

Reviewer 2: The search endpoint will allow the users to find blog items that match the search query.

Reviewer 2: @icode247 since you have reused some of the content from the todo app tutorial, let's make sure there are no references to todo item here in this tutorial.
